_posts/2023-07-24-native-startup-rss-troubleshooting.adoc

```diff
@@ -0,0 +1,228 @@
+
+And here's a screenshot of the mmap calls that relate to heap chunk allocations:
+
+image::rss-before-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.7.6.Final]
+
+Observing the number of times `allocate` is highlighted above, i.e. 2, we can say that the `main` thread in Quarkus 2.7.6.Final allocates roughly 2MB of heap chunks.
+
+For Quarkus 2.13.4.Final, the flamegraph looks like this:
+
+image::rss-after-mmap.svg[startup rss startup flamegraph for Quarkus 2.13.4.Final]
+
+Focusing on the same area, we observe that the `allocate` count is 5 in this case, so the `main` thread allocates roughly 5MB of heap chunks:
+
+image::rss-after-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.13.4.Final]
+
+The increase in object allocation at runtime startup between these two Quarkus versions,
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
+is what was causing the increase in RSS consumption.
+
+Since then, improvements were made to the components that caused the extra object allocation.
+By Quarkus version 2.13.7.Final, the RSS consumption on startup was back to previous levels.
+We are also looking into ways to improve thread local allocation logic in SubstrateVM.
+
+=== `ObjectAllocationInNewTLAB` JFR event
+
+Another improvement since we carried out this investigation can be found in the area of tooling.
+At the time, the only tooling we could really use here was `perf` or `eBPF`, which is only available in Linux.
+Additions to the emitted JFR events for native applications in latest GraalVM Community versions include the `jdk.ObjectAllocationInNewTLAB` event,
+which is the JFR the event that would have highlighted discrepancies in the Quarkus versions compared above.
+Unfortunately, the Quarkus versions tested here are not compatible with this GraalVM version,
+but we can give it a go with the latest Quarkus version to see what comes out.
+
+First, we build the native Quarkus application with JFR support:
+
+[source,bash]
+----
+$ ./mvnw package -DskipTests -Dquarkus.native.monitoring=jfr -Dnative
+----
+
+To obtain `jdk.ObjectAllocationInNewTLAB` events,
+a custom JFC configuration file is required.
+We use `jfr configure` to generate it:
+
+[source,bash]
+----
+$ $JAVA_HOME/bin/jfr configure jdk.ObjectAllocationInNewTLAB#enabled=true --output newtlab.jfc
+----
+
+Then, we start the Quarkus native executable with the necessary JFR flags:
+
+[source,bash]
+----
+$ getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -XX:+FlightRecorder -XX:StartFlightRecording=settings=newtlab.jfc,filename=recording.jfr -Xmx64m
+...
+2023-07-21 12:25:33,739 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 3.2.1.Final) started in 0.019s. Listening on: http://0.0.0.0:8080
+----
+
+After shutting down Quarkus,
+we can use https://adoptium.net/jmc[Adoptium's Eclipse Mission Control],
```

Comment:
Just in case user's are unfamiliar with JMC, it might be good to mention that you can see user-friendly flamegraphs of this JFR event (similar to what you obtained with perf and eBPF).

Reply:
I'm adding a JMC screenshot but I don't think the flamegraphs in JMC add anything productive in this case. It assumes each object allocation is a different flamegraph, because a different object is allocated, and show doesn't show them together in the same way you saw with perf/eBPF.

Reply:
Do you mean you are only able to see the stacktrace of a single event at a time? If you click on the event itself or highlight a series of events, you can view them all together.  Is this what you mean? 
![image](https://github.com/quarkusio/quarkusio.github.io/assets/35396843/9178478b-a3b8-472c-b53e-8401e1346d60)

The behavior you describe (viewing one class at a time) is also possible though:
![image](https://github.com/quarkusio/quarkusio.github.io/assets/35396843/52d13fe5-baa4-4b3b-8d84-1dc1ce488f06)

Reply:
Ah ok. I couldn't get to see that initially. I had to click back on the "Event browser" and then select the event type to see them all together. I'll try to add that to the blog.

Reply:
Yup, I agree the UI is a little clunky. I do wish there was a way to quickly "select all". But as far as I know there isn't :(

Reply:
If you select multiple events at the same time, e.g. all events coming from `main` thread, the flamegraph expands to cover all of those, which as expected all root at the `ApplicationImpl.doStart()` method. I added this screenshot to the blog post.